### PR TITLE
feat: Cassandra on eks

### PR DESCRIPTION
**Pull request by AlyIbrahim**
### What does this PR do?

🛑 Please open an issue first to discuss any significant work and flesh out details/direction - we would hate for your time to be wasted.
Consult the [CONTRIBUTING](https://github.com/awslabs/data-on-eks/blob/main/CONTRIBUTING.md#contributing-via-pull-requests) guide for submitting pull-requests.

Running Cassandra on EKS using K8ssandra Operator

### Motivation

https://github.com/awslabs/data-on-eks/issues/18

### More

- [x] Yes, I have tested the PR using my local account setup (Provide any test evidence report under Additional Notes)
- [x] Mandatory for new blueprints. Yes, I have added a example to support my blueprint PR
- [ ] Mandatory for new blueprints. Yes, I have updated the `website/docs` or `website/blog` section for this feature
- [ ] Yes, I ran `pre-commit run -a` with this PR. Link for installing [pre-commit](https://pre-commit.com/) locally

### For Moderators

- [ ] E2E Test successfully complete before merge?

### Additional Notes

(https://github.com/awslabs/data-on-eks/issues/18)

Working on the docs, but didn't want to hold this PR more
PR is already independent of EKS Bluprints so it compatible with V5

### CODE REVIEW

File: distributed-databases/cassandra/main.tf [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+          ebs = {
+            volume_size = 100
+            volume_type = "gp3"
+          }
+        }
+      }
+
+      labels = {
+        WorkerType    = "ON_DEMAND"
+        NodeGroupType = "cassandra"
```
Comment (vara-bonthu) [inline distributed-databases/cassandra/main.tf:144]: change this to `NodeGroupType = "core"`

File: distributed-databases/cassandra/main.tf [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+      }
+
+      tags = {
+        Name                     = "core-node-grp"
+      }
+    }
+    cassandra_node_group = {
+      name        = "cassandra-node-group"
+      description = "EKS managed node group example launch template"
```
Comment (vara-bonthu) [inline distributed-databases/cassandra/main.tf:153]: May be a meaningful description related to Cassandra

File: distributed-databases/cassandra/main.tf [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+      EOT
+
+      # Optional - Post bootstrap data to verify anything
+      post_bootstrap_user_data = <<-EOT
+        echo "Bootstrap complete.Ready to Go!"
+      EOT
+
+      subnet_ids = module.vpc.private_subnets
+
+      min_size     = 1
```
Comment (vara-bonthu) [inline distributed-databases/cassandra/main.tf:174]: Do we need at least one node running on each az for Cassandra?

File: distributed-databases/cassandra/main.tf [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+  attach_vpc_cni_policy = true
+  vpc_cni_enable_ipv4   = true
+  oidc_providers = {
+    main = {
+      provider_arn               = module.eks.oidc_provider_arn
+      namespace_service_accounts = ["kube-system:aws-node"]
+    }
+  }
+  tags = local.tags
+}
```
Comment (vara-bonthu) [inline distributed-databases/cassandra/main.tf:239]: Move this to addons.tf file once you remove the AWS managed add-ons

File: distributed-databases/cassandra/variables.tf [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+#     aws_partition_id               = string
+#     aws_region_name                = string
+#     eks_cluster_id                 = string
+#     eks_oidc_issuer_url            = string
+#     eks_oidc_provider_arn          = string
+#     tags                           = map(string)
+#     irsa_iam_role_path             = string
+#     irsa_iam_permissions_boundary  = string
+#   })
+# }
```
Comment (vara-bonthu) [inline distributed-databases/cassandra/variables.tf:46]: remove this commented section

File: distributed-databases/cassandra/vpc.tf [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+  name = local.name
+  cidr = local.vpc_cidr
+
+  azs              = local.azs
+  public_subnets   = [for k, v in local.azs : cidrsubnet(local.vpc_cidr, 8, k)]
+  private_subnets  = [for k, v in local.azs : cidrsubnet(local.vpc_cidr, 8, k + 10)]
+  # database_subnets = [for k, v in local.azs : cidrsubnet(local.vpc_cidr, 8, k + 20)]
+
+  # create_database_subnet_group       = true
+  # create_database_subnet_route_table = true
```
Comment (vara-bonthu) [inline distributed-databases/cassandra/vpc.tf:17]: remove these commented lines

File: distributed-databases/cassandra/vpc.tf [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+module "vpc" {
+  source  = "terraform-aws-modules/vpc/aws"
+  version = "~> 3.0"
+
+  name = local.name
+  cidr = local.vpc_cidr
+
+  azs              = local.azs
+  public_subnets   = [for k, v in local.azs : cidrsubnet(local.vpc_cidr, 8, k)]
+  private_subnets  = [for k, v in local.azs : cidrsubnet(local.vpc_cidr, 8, k + 10)]
```
Comment (vara-bonthu) [inline distributed-databases/cassandra/vpc.tf:13]: Lets' use hardcoded private and public subnets. You can use reference from other blueprints.

File: distributed-databases/cassandra/addons.tf [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+#---------------------------------------------------------------
+# Metrics Server
+#---------------------------------------------------------------
+resource "helm_release" "metrics-server" {
+    name        = "metrics-server"
+    description = "The Amazon Elastic Block Store Container Storage Interface (CSI) Driver provides a CSI interface used by Container Orchestrators to manage the lifecycle of Amazon EBS volumes."
+    chart       = "metrics-server"
+    repository  = "https://kubernetes-sigs.github.io/metrics-server/"
+    namespace   = "kube-system"
+    wait        = true
```
Comment (vara-bonthu) [inline distributed-databases/cassandra/addons.tf:31]: this is a default value so you can remove this one

File: distributed-databases/cassandra/addons.tf [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+reclaimPolicy: Delete
+volumeBindingMode: WaitForFirstConsumer
+YAML
+
+  depends_on = [module.eks.cluster_id]
+}
+
+#---------------------------------------------------------------
+# Install K8ssandra Operator
+#---------------------------------------------------------------
```
Comment (vara-bonthu) [inline distributed-databases/cassandra/addons.tf:66]: Incorrect Heading for cert manager

File: distributed-databases/cassandra/addons.tf [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+# Kubernetes Add-ons
+#---------------------------------------------------------------
+
```
Comment (vara-bonthu) [inline distributed-databases/cassandra/addons.tf:4]: Add EKS managed add-ons here

File: distributed-databases/cassandra/addons.tf [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+resource "helm_release" "k8ssandra_operator" {
+  name  = "k8ssandra-operator"
+  chart = "k8ssandra-operator"
+  repository  = "https://helm.k8ssandra.io/stable"
+  version     = "0.39.1"
+  namespace   = "k8ssandra-operator"
+  create_namespace = true
+  description = "K8ssandra Operator to run Cassandra DB on Kubernetes"
+  wait        = true
+  values      = [templatefile("${path.module}/helm-values/values.yaml", {})]
```
Comment (vara-bonthu) [inline distributed-databases/cassandra/addons.tf:107]: rename this to `cassandra-values.yaml`

**Review by vara-bonthu — COMMENTED** [review]
Thanks for the excellent PR @AlyIbrahim  👍🏼 I left few minor comments.

This PR will complete once we we add a doc for this deployment with the best practices(storage, compute etc.) for deploying Cassandra on EKS.

File: distributed-databases/cassandra/main.tf [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+      EOT
+
+      subnet_ids = module.vpc.private_subnets
+
+      min_size     = 1
+      max_size     = 9
+      desired_size = 3
+
+      force_update_version = true
+      instance_types       = ["r6i.2xlarge"]
```
Comment (vara-bonthu) [inline distributed-databases/cassandra/main.tf:179]: change instance type to r6i.8xlarge

File: distributed-databases/cassandra/main.tf [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+
+      ebs_optimized = true
+      block_device_mappings = {
```
Comment (vara-bonthu) [inline distributed-databases/cassandra/main.tf:182]: add comment saying this is a root volume

File: distributed-databases/cassandra/examples/gp3-storage-class.yml [hunk 14]
```diff
@@ -0,0 +1,14 @@
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  annotations:
+    storageclass.kubernetes.io/is-default-class: "true"
+  name: gp3
+parameters:
+  fsType: xfs
+  type: gp3
```
Comment (vara-bonthu) [inline distributed-databases/cassandra/examples/gp3-storage-class.yml:9]: change this IO2. See the above comment

**Comment by askulkarni2** [comment]
Closing due to inactivity.

**Pull request closed without merging**